### PR TITLE
Flow metrics and events

### PR DESCRIPTION
**Pull request by ajbalogh**
[View the proposed changes using the redocly API viewer](https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/open-traffic-generator/models/advanced_metrics/artifacts/openapi.yaml)

A proposal for flow metrics and events. 

This is in response to issue #106  which requests that by default, any configured flow.packet not include additional instrumentation in the payload unless explicitly required.  

Flows by default will not have metrics enabled and the only metrics that can be retrieved by default are port metrics.

In order to get flow metrics a user will have to indicate to the implementation that flow metrics are to be enabled. 
Other flow metric features such as disaggregating by tx/rx port and additional metrics such as timestamps and latency can also be enabled.

```python
import pytest


@pytest.fixture
def primary_link_name():
    return 'Rx Primary Link'


@pytest.fixture
def primary_routes_name():
    return 'Rx Primary Routes'


@pytest.fixture
def secondary_routes_name():
    return 'Rx Secondary Routes'


@pytest.fixture
def config(api, primary_link_name, primary_routes_name, secondary_routes_name):
    """A skeleton configuration suitable for testing flow metrics
    Sets up TX routes and identical RX primary/secondary routes.
    """
    config = api.config()
    p_tx, p_rx_pri, p_rx_sec = config.ports \
        .port(name='Tx Port') \
        .port(name=primary_link_name) \
        .port(name='Secondary Rx Port')
    d_tx, d_rx_pri, d_rx_sec = config.devices \
        .device(name='Tx Device', container_name=p_tx.name) \
        .device(name='Primary Rx Device', container_name=p_rx_pri.name) \
        .device(name='Secondary Rx Device', container_name=p_rx_sec.name)
    tx_rr, = d_tx.ethernet.ipv4.bgpv4.bgpv4_routes.bgpv4route(name='Tx Routes')
    rx_prr, = d_rx_pri.ethernet.ipv4.bgpv4.bgpv4_routes.bgpv4route(name=primary_routes_name)
    rx_srr, = d_rx_sec.ethernet.ipv4.bgpv4.bgpv4_routes.bgpv4route(name=secondary_routes_name)
    flw, = config.flows.flow(name='Tx Ipv4 Device -> Rx Primary/Secondary Routes')
    flw.tx_rx.device.mode = flw.tx_rx.device.ONE_TO_ONE
    flw.tx_rx.device.tx_names = [tx_rr.name]
    flw.tx_rx.device.rx_names = [rx_prr.name, rx_srr.name]
    return config


def test_flow_metrics(api, config):
    """Test configuring flow metrics and retrieving flow metrics
    """
    # configure flow metrics
    flow = config.flows[0]
    flow.packet[0].ethernet.dst.metric_group = 'eth.dst'
    flow.metrics.enable = True
    flow.metrics.ports = True
    flow.metrics.timestamps.enable = True
    flow.metrics.latency.enable = True
    flow.metrics.latency.mode = flow.metrics.latency.STORE_FORWARD
    api.set_config(config)

    # start transmit
    api.set_transmit_state(api.transmit_state())
    
    # retrieve flow metrics
    metric_request = api.metric_request()
    metric_request.flow.flow_names=[config.flows[0].name]
    metric_request.flow.metric_groups.ingress=[flow.packet[0].ethernet.dst.metric_group])
    metric_request.flow.metric_names = []
    metric_response = api.get_metrics(metric_request)
    for flow_metric in metric_response.flow_metrics:
        print(
            flow_metric.name,
            flow_metric.port_tx,
            flow_metric.port_rx,
            flow_metric.metric_groups.name,
            flow_metric.metric_groups.value,
            flow_metric.frames_tx,
            flow_metric.frames_rx,
            flow_metric.latency.average_us,
            flow_metric.timestamps.last_timestamp_us      
        )


if __name__ == '__main__':
    pytest.main(['-vv', '-s', __file__])

```

### CODE REVIEW

File: advanced/config.yaml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+          $ref: '#/components/schemas/Advanced.Config.Event'
+        convergence:
+          $ref: '#/components/schemas/Advanced.Config.Convergence'
+
+    Advanced.Config.Flows:
+      description: |-
+        The optional container for flows metric configuration.
+      type: object
+      properties:
+        enable:
```
Comment (ankur-sheth) [inline advanced/config.yaml:24]: What happens if somebody has not explicitly enabled advanced.config.flows but has enabled advanced.config.latency or advanced.config.timestamps?  Should that request get rejected? Or is it implementation dependent?

File: advanced/event.yaml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+          type: string
+          x-constraint:
+          - '#/components/schemas/Port/properties/name'
+          - '#/components/schemas/Flow/properties/name'
+          - '#/components/schemas/Device.Bgpv4Route/properties/name'
+          - '#/components/schemas/Device.Bgpv6Route/properties/name'
+          example: Bgpv4 Route Range 1
+        begin_timestamp_ns:
+          description: |-
+            The time in seconds at which the event began
```
Comment (ankur-sheth) [inline advanced/event.yaml:64]: Is this in seconds or nanoseconds?

File: advanced/event.yaml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+          type: number
+          format: double
+        end_timestamp_ns:
+          description: |-
+            The time in seconds at which the event ended
```
Comment (ankur-sheth) [inline advanced/event.yaml:69]: Is this in seconds or nanoseconds?

File: api/api.yaml [hunk 4]
```diff
@@ -1,25 +1,7 @@
...
-  x-model-guide: https://github.com/open-traffic-generator/models/blob/master/MODEL-GUIDE.md
-
-servers: 
-  - url: /
-
 paths: 
   /config:
     post:
-      tags: [Configuration]
+      tags: ['Traffic Generator']
```
Comment (ankur-sheth) [inline api/api.yaml:4]: What is `tag` used for?

File: api/api.yaml [hunk 5]
```diff
@@ -161,26 +163,11 @@ paths:
               schema:
                 $ref: '../result/state.yaml#/components/schemas/State.Metrics'
 
-  /results/capabilities:
```
Comment (ankur-sheth) [inline api/api.yaml:164]: Are we getting rid of the capabilities API?  There are references to the capabilities API in `capture/capture.yaml`, `layer1/layer1.yaml`, `port/port.yaml` not to mention `result/capability.yaml`.  If we want to get rid of it, we should do it in all these places & delete result/capability.yaml but do it in a separate PR.

File: api/advanced-api.yaml [hunk 6]
```diff
@@ -0,0 +1,114 @@
+paths:
+  /advanced/config:
+    post:
+      tags: ['Advanced']
```
Comment (ankur-sheth) [inline api/advanced-api.yaml:4]: Should the tag be 'Advanced Configuration'?

File: api/advanced-api.yaml [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        '409':
+          $ref: '../result/request.yaml#/components/responses/Conflict'
+        '590':
+          $ref: '../result/request.yaml#/components/responses/Fail'
+        '591':
+          $ref: '../result/request.yaml#/components/responses/NotSupported'
+
+  /advanced/metrics:
+    post:
+      tags: ['Advanced']
```
Comment (ankur-sheth) [inline api/advanced-api.yaml:27]: Should the tag be 'Advanced Metrics'?

File: api/advanced-api.yaml [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        '409':
+          $ref: '../result/request.yaml#/components/responses/Conflict'
+        '590':
+          $ref: '../result/request.yaml#/components/responses/Fail'
+        '591':
+          $ref: '../result/request.yaml#/components/responses/NotSupported'
+
+  /advanced/analytics:
+    post:
+      tags: ['Advanced']
```
Comment (ankur-sheth) [inline api/advanced-api.yaml:54]: Should this tag be 'Advanced Analytics'?

File: api/advanced-api.yaml [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        '409':
+          $ref: '../result/request.yaml#/components/responses/Conflict'
+        '590':
+          $ref: '../result/request.yaml#/components/responses/Fail'
+        '591':
+          $ref: '../result/request.yaml#/components/responses/NotSupported'
+
+  /event/subscribe:
+    post:
+      tags: ['Advanced']
```
Comment (ankur-sheth) [inline api/advanced-api.yaml:81]: Should the tag be 'Advanced Events' or something like that?

**Review by ankur-sheth — CHANGES_REQUESTED** [review]

File: api/info.yaml [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  title: Open Traffic Generator API
+  description: |-
+    The Open Traffic Generator API consists of the following features:
+    - a common Traffic Generator API
+    - an Advanced Metrics API feature (under review)
+
+    Contributions can be made in the following ways:
+    - [open an issue](https://github.com/open-traffic-generator/models/issues) in the models repository
+    - [fork the models repository](https://github.com/open-traffic-generator/models) and submit a PR
+  version: 0.3.6
```
[the remark this answers is not in the document]
Comment (ajbalogh) [reply api/info.yaml:13]: will move to 0.4.0

File: advanced/config.yaml [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+          description: |-
+            True to enable metrics for configured flows.
+            Enabling this option could possibly change the packet payload as 
+            implementations may add instrumentation values to identify which 
+            flow a packet is associated with.
+          type: boolean
+          default: false
+
+    Advanced.Config.Timestamp:
```
[the remark this answers is not in the document]
Comment (ajbalogh) [reply advanced/config.yaml:33]: will remove timestamp option

File: advanced/config.yaml [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        The optional container for flow timestamp metric configuration.
+      type: object
+      properties:
+        enable:
+          description: |-
+            True to enable timestamp metrics. 
+            This adds first and last timestamp metrics to individual flows.
+          type: boolean
+          default: false
+        synchronization:
```
[the remark this answers is not in the document]
Comment (ajbalogh) [reply advanced/config.yaml:44]: no use cases so removing for now

File: advanced/config.yaml [hunk 13]
```diff
@@ -0,0 +1,114 @@
+components:
+  schemas:
+    Advanced.Config:
+      description: |-
+        The container for the configuration of advanced features.
+      type: object
+      properties:
+        flows:
```
[the remark this answers is not in the document]
Comment (ajbalogh) [reply advanced/config.yaml:8]: OpenAPI properties by default are optional unless specified by using the `required` keyword.  This means by default all of the advanced options are not enabled.

File: advanced/config.yaml [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+          description: |-
+            True to enable events. 
+          type: boolean
+          default: false
+        rx_rate_threshold:
+          description: |-
+            An event will be generated when the rx rate passes
+            above or below this value.
+          type: number
+          maximum: 100
```
[the remark this answers is not in the document]
Comment (ajbalogh) [reply advanced/config.yaml:111]: Setting minimum to 0

File: api/api.yaml (hunk 5 above)
Comment (ajbalogh) [reply api/api.yaml:164]: Capabilities will be removed and will be done in a separate PR>

File: api/advanced-api.yaml (hunk 6 above)
Comment (ajbalogh) [reply api/advanced-api.yaml:4]: Keeping all advanced APIs under the tag `Advanced` rather than two separate tags for config and metrics.  This can easily be changed in the future without incurring any compatibility issues.

File: api/advanced-api.yaml (hunk 7 above)
Comment (ajbalogh) [reply api/advanced-api.yaml:27]: Keeping all advanced APIs under the tag Advanced rather than two separate tags for config and metrics. This can easily be changed in the future without incurring any compatibility issues.

File: api/advanced-api.yaml (hunk 8 above)
Comment (ajbalogh) [reply api/advanced-api.yaml:54]: Keeping all advanced APIs under the tag Advanced rather than two separate tags for config and metrics. This can easily be changed in the future without incurring any compatibility issues.

File: api/advanced-api.yaml (hunk 9 above)
Comment (ajbalogh) [reply api/advanced-api.yaml:81]: Keeping all advanced APIs under the tag Advanced rather than two separate tags for config and metrics. This can easily be changed in the future without incurring any compatibility issues.

File: advanced/config.yaml (hunk 1 above)
Comment (ajbalogh) [reply advanced/config.yaml:24]: That will be implementation independent

File: advanced/event.yaml (hunk 3 above)
Comment (ajbalogh) [reply advanced/event.yaml:69]: Fixing description to reflect nanoseconds.

File: api/api.yaml (hunk 4 above)
Comment (ajbalogh) [reply api/api.yaml:4]: From OpenAPI docs...
> A list of tags for API documentation control. Tags can be used for logical grouping of operations by resources or any other qualifier.

File: advanced/event.yaml (hunk 2 above)
Comment (ajbalogh) [reply advanced/event.yaml:64]: Fixing description to reflect nanoseconds

**Comment by ashutshkumr** [comment]
We need to expose 590 response for other APIs as well - e.g. get_metrics, get_capture, get_config

File: advanced/config.yaml [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+    Advanced.Config.Latency:
+      description: |-
+        The optional container for latency metric configuration.
+      type: object
+      properties:
+        enable:
+          description: |-
+            True to enable latency metrics using timestamps.
+            Enabling this option could possibly change flow packet payloads as
```
Comment (ashutshkumr) [inline advanced/config.yaml:23]: Enabling this option requires enabling flow metrics which may affect the resultant packet payload due to additional instrumentation data.

File: advanced/config.yaml [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+            implementations may add instrumentation values to the flow
+            packet payload.
+          type: boolean
+          default: false
+
+    Advanced.Config.Event:
+      description: |-
+        The optional container for event configuration.
+      type: object
+      properties:
```
Comment (ashutshkumr) [inline advanced/config.yaml:69]: Shouldn't we provide separate containers for separate events per https://github.com/open-traffic-generator/models/pull/95#discussion_r613108350 ?

Comment (ashutshkumr) [reply advanced/config.yaml:69]: Also, not all events are tied to flow metrics (instrumentation signature).

File: advanced/config.yaml [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+            on the receiving port (FIFO).  This is based on the RFC 1242 
+            standard.
+          type: string
+          enum: 
+          - store_forward
+          - cut_through
+          default: store_forward
+
+    Advanced.Config.Convergence:
+      description: |-
```
Comment (ashutshkumr) [inline advanced/config.yaml:52]: It is not clear from the description that enabling this requires users to also set rate threshold in events container.

**Review by ankur-sheth — COMMENTED** [review]
LGTM

File: advanced/analytics.yaml [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+        data_plane_convergence_ns:
+          description: |-
+            The convergence time measured from the data plane perspective only. 
+            It measures the time from BTT, when the rate on Test Port 2 crosses 
+            below the Rx Threshold until an acceptable amount of traffic was 
+            received at time ATT, when the rate crosses above the configured 
+            rx_rate_threshold.
+          type: number
+          format: double
+        control_plane_data_plane_convergence_ns:
```
Comment (ANISH-GOTTAPU) [inline advanced/analytics.yaml:71]: in our internal discussion, we felt microseconds unit is better than nanoseconds if it is huge value. please provide your inputs

Adding snippet from ixn
image

File: advanced/metrics.yaml [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+          $ref: '#/components/schemas/AdvancedMetrics.Timestamp'
+        latency:
+          $ref: '#/components/schemas/AdvancedMetrics.Latency'
+
+    AdvancedMetrics.Timestamp:
+      description: |-
+        The container for timestamp metrics.
+      type: object
+      properties:
+        first_timestamp_ns:
```
Comment (ANISH-GOTTAPU) [inline advanced/metrics.yaml:123]: can we have timestamp in this format (hours/minutes/seconds) instead of ns?

Comment (ashutshkumr) [reply advanced/metrics.yaml:123]: I think we should stick to unix timestamps, otherwise a date-string may further complicate the script.

**Review by ankur-sheth — APPROVED** [review]

File: control/control.yaml [hunk 20]
```diff
@@ -65,50 +65,31 @@ components:
...
-            The names of Bgpv4RouteRange object to control.
+            The names of device route objects to control.
+            If no names are specified then all route objects will be affected.
           type: array
           items:
             type: string
-            x-constraint:
-            - '/components/schemas/Device.Bgpv4RouteRange/properties/name'
+          x-constraint:
+          - '/components/schemas/Device.Bgp.Route/properties/name'
```
Comment (ANISH-GOTTAPU) [inline control/control.yaml:90]: `Device.Bgp.Route` node is not available, because of which `set_route_state`  is not getting generated in snappi

File: result/flow.yaml [hunk 21]
```diff
@@ -1,97 +1,116 @@
...
 components:
   schemas:
     Flow.Metrics.Request:
       description: >-
-        The request to the traffic generator for flow results.
+        The request container for flow metrics.
       type: object
       properties:
-        flow_names:
+        flow_groups:
```
Comment (ANISH-GOTTAPU) [inline result/flow.yaml:8]: with `flow_groups` , user side complexity will increase. If I have 10 flows and want to retrieve 5 flows, have to create 5 groups.

Ex:
    request = api.metrics_request()
    f1, f2, f3, f4, f5 = request.flow.flowgroups.group().group().group().group().group()
    f1.flow_name = 'f1'
    f2.flow_name = 'f2'
    f3.flow_name = 'f3'
    f4.flow_name = 'f4'
    f5.flow_name = 'f5'
    return api.get_flow_metrics(request)

File: result/flow.yaml [hunk 22]
```diff
@@ -116,28 +135,49 @@ components:
...
+          $ref: '#/components/schemas/Metric.Timestamp'
+        latency:
+          $ref: '#/components/schemas/Metric.Latency'
+
+    Metric.Timestamp:
+      description: |-
+        The container for timestamp metrics.
+      type: object
+      properties:
+        first_timestamp_us:
```
Comment (ANISH-GOTTAPU) [inline result/flow.yaml:152]: Based on our previous discussion, all timestamps can be `nano`

File: result/flow.yaml (hunk 21 above)
Comment (ajbalogh) [reply result/flow.yaml:8]: aab67e5

File: control/control.yaml (hunk 20 above)
Comment (ajbalogh) [reply control/control.yaml]: aab67e5

**Comment by ashutshkumr** [comment]
Couple of more remarks from local discussion (list might change a bit):
- remove flow.metrics.port from config, since nobody asked for it
- flow metric query will have flow_names in the top level, as it used to be before
- flow metric query will have a choice on whether we want to fetch results based on ingress or egress
- flow metrics will always have port_tx and port_rx
- flow metrics will also have ingress and egress groups at the top level, hence no groups parameter

File: result/flow.yaml (hunk 22 above)
Comment (ajbalogh) [reply result/flow.yaml]: d9ba3bb

**Review by ashutshkumr — APPROVED** [review]
